Community login support

leog opened the pull request:
## What does this PR do?

Allow a discourse instance accompanying Cal.com instance to login using Cal.com login. 

The login to Cal Community from Cal.com:
 - Reuses username (not editable from community)
 - Reuses name (not editable from community)
 - Reuses avatar (not editable from community)
 - Defines user's profile website field as Cal.com/username
 - Reuses role
 - Reuses bio

Pending:
 - [X] ~Test and make work non-credential login such as Google or SAML~ Can't be done as of now, callbackUrl gets missed between logging in with Google and the signIn method that could work out a community login redirect. For that reason, non-credential providers are hidden now when a community login is detected.
 - [X] ~Test changing username/email to see if user in Community also changes username/email keeping posts and other activity unchanged/not resetting~ Not needed, `external_id` is constant and it is what the Community system relies on
 - [x] Copy color scheme between dark and light to make it look always the same (dark), notice the button or info blocks in light now using some violet

Before merging, to get it to work on PROD, do the following:
 - [ ] Add COMMUNITY_SECRET env var in Vercel to match Discourse Connect secret field
 - [ ] Change Discourse Connect url field to http://app.cal.com/auth/login
 - [ ] Delete all testing accounts created in https://Cal.Community

**Environment**: Staging(main branch) / Production

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How should this be tested?

Go to https://cal.community and sign in using any valid account

leog committed:
Community login support

leog committed:
Community secret as env var

leog committed:
Leave out self-hosters for now

leog commented on apps/web/package.json:
```diff
@@ -69,6 +69,7 @@
     "async": "^3.2.4",
     "bcryptjs": "^2.4.3",
     "classnames": "^2.3.1",
+    "discourse-sso": "^1.0.5",
```
Using this dependency to do the heavy lifting behind the logic to verify a login request from Discourse.

leog commented on apps/web/package.json:
```diff
@@ -137,6 +138,7 @@
     "@types/async": "^3.2.15",
     "@types/bcryptjs": "^2.4.2",
     "@types/detect-port": "^1.3.2",
+    "@types/discourse-sso": "^1.0.1",
```
Types for the newly added dependency come separately.

leog commented on apps/web/pages/api/auth/[...nextauth].tsx:
```diff
@@ -62,6 +70,7 @@ const providers: Provider[] = [
           role: true,
           id: true,
           username: true,
+          bio: true,
```
Adding bio to pass it down to Discourse, we can remove it if we don't want it.

leog commented on apps/web/pages/api/auth/[...nextauth].tsx:
```diff
@@ -248,6 +259,39 @@ function isNumber(n: string) {
 
 const calcomAdapter = CalComAdapter(prisma);
 
+const checkCommunityLogin = (
```
All the logic to handle Discourse sign in from our current login page.

leog commented on apps/web/pages/api/auth/[...nextauth].tsx:
```diff
@@ -248,6 +259,39 @@ function isNumber(n: string) {
+  const params = new URLSearchParams(callbackUrl.searchParams);
+  const ssoParam = params.get("sso");
+  const sigParam = params.get("sig");
+  const sso = new discourseSso(process.env.COMMUNITY_SECRET);
```
This is the shared secret between our app and Discourse to avoid anyone messing our instance.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -36,6 +37,12 @@ interface LoginValues {
+  sig: z.string().optional(),
+  callbackUrl: z.string().optional().default(""),
+});
+
```
Converting already existing query parameters into Zod schema, adding `sso` and `sig` for Discourse.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -51,6 +58,7 @@ export default function Login({
 
   const [twoFactorRequired, setTwoFactorRequired] = useState(false);
   const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const [successMessage, setSuccessMessage] = useState<string | null>(null);
```
It's always nice to show a success message to the user once everything went well and a redirection is about to happen.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -59,22 +67,23 @@ export default function Login({
     [ErrorCode.IncorrectTwoFactorCode]: `${t("incorrect_2fa_code")} ${t("please_try_again")}`,
     [ErrorCode.InternalServerError]: `${t("something_went_wrong")} ${t("please_try_again_and_contact_us")}`,
     [ErrorCode.ThirdPartyIdentityProviderEnabled]: t("account_created_with_identity_provider"),
+    [ErrorCode.InvalidCommunityLogin]: `${t("something_went_wrong")} ${t("signing_in_community")}`,
```
Anything that can happen from the community login will result in this error.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -59,22 +67,23 @@ export default function Login({
+  const safeCallbackUrl = getSafeRedirectUrl(queryParams.callbackUrl);
 
-  callbackUrl = safeCallbackUrl || "";
+  queryParams.callbackUrl = safeCallbackUrl || "";
```
Above lines is just moving from simple query checking to Zod schema, nothing changed really in terms of logic.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -94,17 +103,22 @@ export default function Login({
     </Button>
   );
 
+  const communityLogin = queryParams.sso !== undefined && queryParams.sig !== undefined;
```
We need to be able to know if there is an attempt to login to community to change a bit the sign in process below.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -94,17 +103,22 @@ export default function Login({
+      callbackUrl: communityLogin
+        ? `${COMMUNITY_URL}?sso=${queryParams.sso}&sig=${queryParams.sig}`
+        : queryParams.callbackUrl,
+      redirect: communityLogin,
```
If we are dealing with a community login, then we need to enable redirection and provide the callbackUrl to let the sign in process know we are dealing with that community login.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -94,17 +103,22 @@ export default function Login({
+        : queryParams.callbackUrl,
+      redirect: communityLogin,
     });
+    if (!res && communityLogin) return setSuccessMessage("Success! Redirecting...");
```
When community login succeeds, then no `res` is returned and when that happens in the context of a community login, everything went well, so showing message while being redirected.

leog replied:
Oh I guess this string needs to be localized 😅

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -94,17 +103,22 @@ export default function Login({
     if (!res) setErrorMessage(errorMessages[ErrorCode.InternalServerError]);
     // we're logged in! let's do a hard refresh to the desired url
-    else if (!res.error) router.push(callbackUrl);
+    else if (!res.error) router.push(queryParams.callbackUrl);
```
Relying on Zod parsed query parameters now.

leog commented on apps/web/pages/auth/login.tsx:
```diff
@@ -162,6 +176,7 @@ export default function Login({
               {twoFactorRequired && <TwoFactor center />}
 
               {errorMessage && <Alert severity="error" title={errorMessage} />}
+              {successMessage && <Alert severity="info" title={successMessage} />}
```
For some reason the severity "success" was not what I expected in terms of color and style, so using "info" for now.

PeerRich commented:
the open question still stands if we want discourse code in the core code base cc @zomars

leog commented:
> the open question still stands if we want discourse code in the core code base cc @zomars

Would it help if I get rid of any reference to discourse from the code? it will be just references to community

PeerRich commented:
wasnt referring to the naming just the extra code and dependencies being added. I have no strong opinion, will defer to @zomars

leog committed:
Relying on existant dependencies, moved logic

leog commented:
> wasnt referring to the naming just the extra code and dependencies being added. I have no strong opinion, will defer to @zomars

> @PeerRich gotcha; moved things to `lib/auth` and removed discourse new dependency, now I'm using dependencies we already had.

PeerRich commented:
ive changed my mind-- i dont mind having community log in code here as long as it works 🙏

leog committed:
Adjusting login options and logic

leog committed:
Merge branch 'main' into feat/community-login-support

deepakprabhakara reviewed (approved):
LGTM, nice work @leog

zomars reviewed (changes requested):
Let's move this to [@calcom/auth](https://github.com/calcom/auth) 🙏🏽

zomars committed:
Merge branch 'main' into feat/community-login-support

# Conflicts:
#	apps/web/pages/api/auth/[...nextauth].tsx
#	apps/web/pages/auth/login.tsx

hariombalhara commented:
Marking it draft based on @zomars last comment.

leog commented:
@hariombalhara better to close it, we want to do this in a different package, so we could reuse the logic but this PR is no longer valid.
